discord/ext/commands/parameters.py

```diff
@@ -0,0 +1,269 @@
+        if the parameter has no default.
+        """
+        if self.annotation is empty:
+            return type(self.default) if self.default not in (empty, None) else str
+
+        return self.annotation
+
+    @property
+    def displayed_default(self) -> str | None:
+        """The displayed default in :class:`Command.signature`."""
+        if self._displayed_default is not empty:
+            return self._displayed_default
+
+        return None if self.required else str(self.default)
+
+    async def get_default(self, ctx: Context) -> Any:
+        """|coro|
+
+        Gets this parameter's default value.
+
+        Parameters
+        ----------
+        ctx: :class:`Context`
+            The invocation context to get the default argument for.
+        """
+        # pre-condition: required is False
+        if callable(self.default):
+            return await maybe_coroutine(self.default, ctx)  # type: ignore
+        return self.default
+
+
+def parameter(
+    *,
+    converter: Any = empty,
+    default: Any = empty,
+    displayed_default: str = empty,
+) -> Any:
+    r"""parameter(\*, converter=..., default=..., displayed_default=...)
+
+    A way to assign custom metadata for a :class:`Command`\'s parameter.
+
+    Examples
+    --------
+
+    A custom converter, whilst annotating a parameter with a custom converter works at runtime, type checkers don't like it
+    cause they can't understand what's going on.
+
+    .. code-block:: python3
+
+        class SomeType:
+            foo: int
+
+        class MyVeryCoolConverter(commands.Converter[SomeType]):
+            ...  # implementation left as an exercise for the reader
+
+        @bot.command()
+        async def bar(ctx, cool_value: MyVeryCoolConverter):
+            cool_value.foo  # type checker warns MyVeryCoolConverter has no value foo (uh-oh)
+
+    However, fear not we can use :func:`parameter` to tell type checkers what's going on.
+
+    .. code-block:: python3
+
+        @bot.command()
+        async def bar(ctx, cool_value: SomeType = commands.parameter(converter=MyVeryCoolConverter)):
+            cool_value.foo  # no error (hurray)
+
+    A custom default can be used to have late binding behaviour
+
+    .. code-block:: python3
+
+        @bot.command()
+        async def wave(to: discord.User = commands.parameter(default=lambda ctx: ctx.author)):
+            await ctx.send(f'Hello {to.mention} :wave:')
+
+    Because this is such a common use-case, the library provides :obj:`.Author`, :obj:`.CurrentChannel` and
+    :obj:`.CurrentGuild`, armed with this we can simplify ``wave`` to:
+
+    .. code-block:: python3
+
+        @bot.command()
+        async def wave(to: discord.User = commands.Author):
+            await ctx.send(f'Hello {to.mention} :wave:')
+
+    :obj:`.Author` and co also have other benefits like having the displayed default being filled.
```

Comment:
I think this belongs in the prose docs instead (commands.rst) and an example can be added here.